### ISSUE TITLE
sudo_user in task should accept variables defined outside playbook

### ISSUE THREAD

**Issue by leucos** [ansible/ansible#3048]
This is a followup for #1665 
Variables defined outside playbooks are currently not expanded in the `sudo_user` context, while variables defined in a `vars` section are.

Here is a setup to reproduce the problem :
- inventory file :

```
# hosts file
test
```
- host vars :

```
# host_vars/test
complex:
  user: complexfoo

simple_user: simplebar
```
- playbook (real life : I used 3 different pbooks) :

``` yaml
- name: Tying with simple var
  hosts: all
  sudo_user: ${simple_user}
  connection: local
  tasks:
    debug: msg=${simple_user}
    ignore_errors: yes

- name: Tying with complex var
  hosts: all
  sudo_user: ${complex.user}
  connection: local
  tasks:
    debug: msg="Using ${complex.user}"

- name: Tying with simple inline var
  hosts: all
  vars:
    simple_user: leucos
  sudo_user: ${simple_user}
  connection: local
  tasks:
    debug: msg="Using ${simple_user}"
```

The ouput is :

```
PLAY [Tying with simple var] ************************************************** 

GATHERING FACTS *************************************************************** 
<test> EXEC ['/bin/sh', '-c', 'mkdir -p /tmp/ansible-1369816833.04-56889184067465 && chmod a+rx /tmp/ansible-1369816833.04-56889184067465 && echo /tmp/ansible-1369816833.04-56889184067465']
<test> REMOTE_MODULE setup 
<test> PUT /tmp/tmp3EBvjL TO /tmp/ansible-1369816833.04-56889184067465/setup
<test> EXEC ['/bin/sh', '-c', 'chmod a+r /tmp/ansible-1369816833.04-56889184067465/setup']
<test> EXEC /bin/sh -c 'sudo -k && sudo -H -S -p "[sudo via ansible, key=mnprcnlijfunmhajvbteegjvtjujtvxi] password: " -u ${simple_user} /bin/sh -c '"'"'python2 /tmp/ansible-1369816833.04-56889184067465/setup'"'"''
<test> EXEC ['/bin/sh', '-c', 'rm -rf /tmp/ansible-1369816833.04-56889184067465/ >/dev/null 2>&1']
failed: [test] => {"failed": true, "parsed": false}
invalid output was: usage: sudo [-D level] -h | -K | -k | -V
usage: sudo -v [-AknS] [-D level] [-g groupname|#gid] [-p prompt] [-u user
            name|#uid]
usage: sudo -l[l] [-AknS] [-D level] [-g groupname|#gid] [-p prompt] [-U user
            name] [-u user name|#uid] [-g groupname|#gid] [command]
usage: sudo [-AbEHknPS] [-C fd] [-D level] [-g groupname|#gid] [-p prompt] [-u
            user name|#uid] [-g groupname|#gid] [VAR=value] [-i|-s] [<command>]
usage: sudo -e [-AknS] [-C fd] [-D level] [-g groupname|#gid] [-p prompt] [-u
            user name|#uid] file ...


PLAY RECAP ******************************************************************** 
           to retry, use: --limit @/var/tmp/ansible/pb_simple.retry

test                       : ok=0    changed=0    unreachable=0    failed=1   


PLAY [Tying with complex var] ************************************************* 

GATHERING FACTS *************************************************************** 
<test> EXEC ['/bin/sh', '-c', 'mkdir -p /tmp/ansible-1369816833.54-180309236265737 && chmod a+rx /tmp/ansible-1369816833.54-180309236265737 && echo /tmp/ansible-1369816833.54-180309236265737']
<test> REMOTE_MODULE setup 
<test> PUT /tmp/tmp2QYDhj TO /tmp/ansible-1369816833.54-180309236265737/setup
<test> EXEC ['/bin/sh', '-c', 'chmod a+r /tmp/ansible-1369816833.54-180309236265737/setup']
<test> EXEC /bin/sh -c 'sudo -k && sudo -H -S -p "[sudo via ansible, key=sxryxawxepauvyaxchdhgccctsfwnzzv] password: " -u ${complex.user} /bin/sh -c '"'"'python2 /tmp/ansible-1369816833.54-180309236265737/setup'"'"''
<test> EXEC ['/bin/sh', '-c', 'rm -rf /tmp/ansible-1369816833.54-180309236265737/ >/dev/null 2>&1']
failed: [test] => {"failed": true, "parsed": false}
invalid output was: /bin/sh: ${complex.user} : mauvaise substitution


PLAY RECAP ******************************************************************** 
           to retry, use: --limit @/var/tmp/ansible/pb_complex.retry

test                       : ok=0    changed=0    unreachable=0    failed=1   


PLAY [Tying with simple inline var] ******************************************* 

GATHERING FACTS *************************************************************** 
<test> EXEC ['/bin/sh', '-c', 'mkdir -p /tmp/ansible-1369816834.02-151673826209778 && chmod a+rx /tmp/ansible-1369816834.02-151673826209778 && echo /tmp/ansible-1369816834.02-151673826209778']
<test> REMOTE_MODULE setup 
<test> PUT /tmp/tmpq2PToj TO /tmp/ansible-1369816834.02-151673826209778/setup
<test> EXEC ['/bin/sh', '-c', 'chmod a+r /tmp/ansible-1369816834.02-151673826209778/setup']
<test> EXEC /bin/sh -c 'sudo -k && sudo -H -S -p "[sudo via ansible, key=zwddmrjmpoidlgbhlzqpszkkkhubfghi] password: " -u leucos /bin/sh -c '"'"'python2 /tmp/ansible-1369816834.02-151673826209778/setup'"'"''
<test> EXEC ['/bin/sh', '-c', 'rm -rf /tmp/ansible-1369816834.02-151673826209778/ >/dev/null 2>&1']
ok: [test]

PLAY RECAP ******************************************************************** 
test                       : ok=1    changed=0    unreachable=0    failed=0   
```

**Comment by leucos** [comment]
Mmmmhhh... my bad. Syntax was wrong. It seems to work now.

**Comment by leucos** [comment]
For reference, using checkout, it woks properly for me now with the setup above but with a proper playbook :

```
- name: Tying with simple var
  hosts: all
  gather_facts: no
  sudo_user: ${simple_user}
  connection: local
  tasks:
    - debug: msg=${simple_user}
- name: Tying with complex var
  hosts: all
  gather_facts: no
  sudo_user: ${complex.user}
  connection: local
  tasks:
    - debug: msg="Using ${complex.user}"

- name: Tying with simple inline var
  hosts: all
  vars:
    simple_user: leucos
  sudo_user: ${simple_user}
  connection: local
  tasks:
    - debug: msg="Using ${simple_user}"
```

Output : 

```
PLAY [Tying with simple var] ************************************************** 

TASK: [debug msg=${simple_user}] ********************************************** 
ok: [test] => {"msg": "simplebar"}

PLAY RECAP ******************************************************************** 
test                       : ok=1    changed=0    unreachable=0    failed=0   


PLAY [Tying with complex var] ************************************************* 

TASK: [debug msg="Using ${complex.user}"] ************************************* 
ok: [test] => {"msg": "Using complexfoo"}

PLAY RECAP ******************************************************************** 
test                       : ok=1    changed=0    unreachable=0    failed=0   


PLAY [Tying with simple inline var] ******************************************* 

TASK: [debug msg="Using leucos"] ********************************************** 
ok: [test] => {"msg": "Using leucos"}

PLAY RECAP ******************************************************************** 
test                       : ok=2    changed=0    unreachable=0    failed=0   
```

May be @sfllaw @steverweber and @akerl can confirm ?

**Comment by leucos** [comment]
After few more tries and coffee cup, I can confirm that the bug is still here : **sudo_user in task does not accept variables defined outside playbook**.

Here is an example :

``` yaml
# hosts file
test ansible_ssh_host=put.a.real.hostname.here
```
- host vars :

``` yaml
# host_vars/test
complex:
  user: hadrien

simple_user: hadrien
```
- playbook :

``` yaml
- name: Tying with simple inline var
  hosts: all
  gather_facts: no
  vars:
    inline_simple_user: hadrien
  sudo_user: ${inline_simple_user}
  sudo: yes
  tasks:
    - shell: whoami
      ignore_errors: yes

- name: Tying with complex var
  hosts: all
  gather_facts: no
  sudo_user: ${complex.user}
  sudo: yes
  tasks:
    - shell: whoami
      ignore_errors: yes

- name: Tying with simple var
  hosts: all
  gather_facts: no
  sudo_user: ${simple_user}
  sudo: yes
  tasks:
    - shell: whoami
      ignore_errors: yes
```

Executing the playbook gives :

```
PLAY [Tying with simple inline var] ******************************************* 

TASK: [shell whoami] ********************************************************** 
<test> ESTABLISH CONNECTION FOR USER: root on PORT 22 TO test
<test> EXEC /bin/sh -c 'mkdir -p /tmp/ansible-1370517035.79-68677943528285 && chmod a+rx /tmp/ansible-1370517035.79-68677943528285 && echo /tmp/ansible-1370517035.79-68677943528285'
<test> REMOTE_MODULE command whoami #USE_SHELL
<test> PUT /tmp/tmpuxvkC6 TO /tmp/ansible-1370517035.79-68677943528285/command
<test> EXEC /bin/sh -c 'chmod a+r /tmp/ansible-1370517035.79-68677943528285/command'
<test> EXEC /bin/sh -c 'sudo -k && sudo -H -S -p "[sudo via ansible, key=rzzobxrspxjwejehqqyezaxudqmwezvt] password: " -u hadrien /bin/sh -c '"'"'/usr/bin/python /tmp/ansible-1370517035.79-68677943528285/command'"'"''
<test> EXEC /bin/sh -c 'rm -rf /tmp/ansible-1370517035.79-68677943528285/ >/dev/null 2>&1'
changed: [test] => {"changed": true, "cmd": "whoami ", "delta": "0:00:00.003237", "end": "2013-06-06 13:10:36.301141", "rc": 0, "start": "2013-06-06 13:10:36.297904", "stderr": "", "stdout": "hadrien"}

PLAY [Tying with complex var] ************************************************* 

TASK: [shell whoami] ********************************************************** 
<test> ESTABLISH CONNECTION FOR USER: root on PORT 22 TO test
<test> EXEC /bin/sh -c 'mkdir -p /tmp/ansible-1370517036.68-224153666200229 && chmod a+rx /tmp/ansible-1370517036.68-224153666200229 && echo /tmp/ansible-1370517036.68-224153666200229'
<test> REMOTE_MODULE command whoami #USE_SHELL
<test> PUT /tmp/tmpDRc_Es TO /tmp/ansible-1370517036.68-224153666200229/command
<test> EXEC /bin/sh -c 'chmod a+r /tmp/ansible-1370517036.68-224153666200229/command'
<test> EXEC /bin/sh -c 'sudo -k && sudo -H -S -p "[sudo via ansible, key=trfozilptlahitjpwrrpjsezlckfavmx] password: " -u ${complex.user} /bin/sh -c '"'"'/usr/bin/python /tmp/ansible-1370517036.68-224153666200229/command'"'"''
<test> EXEC /bin/sh -c 'rm -rf /tmp/ansible-1370517036.68-224153666200229/ >/dev/null 2>&1'
failed: [test] => {"failed": true, "parsed": false}
invalid output was: /bin/sh: 1: Bad substitution

...ignoring

PLAY [Tying with simple var] ************************************************** 

TASK: [shell whoami] ********************************************************** 
<test> ESTABLISH CONNECTION FOR USER: root on PORT 22 TO test
<test> EXEC /bin/sh -c 'mkdir -p /tmp/ansible-1370517037.4-31942410728002 && chmod a+rx /tmp/ansible-1370517037.4-31942410728002 && echo /tmp/ansible-1370517037.4-31942410728002'
<test> REMOTE_MODULE command whoami #USE_SHELL
<test> PUT /tmp/tmp0CsmQ6 TO /tmp/ansible-1370517037.4-31942410728002/command
<test> EXEC /bin/sh -c 'chmod a+r /tmp/ansible-1370517037.4-31942410728002/command'
<test> EXEC /bin/sh -c 'sudo -k && sudo -H -S -p "[sudo via ansible, key=hifajiaqrdcxefvcdjsdyqungntiisjw] password: " -u ${simple_user} /bin/sh -c '"'"'/usr/bin/python /tmp/ansible-1370517037.4-31942410728002/command'"'"''
<test> EXEC /bin/sh -c 'rm -rf /tmp/ansible-1370517037.4-31942410728002/ >/dev/null 2>&1'
fatal: [test] => failed to parse: usage: sudo [-D level] -h | -K | -k | -V
usage: sudo -v [-AknS] [-D level] [-g groupname|#gid] [-p prompt] [-u user
            name|#uid]
usage: sudo -l[l] [-AknS] [-D level] [-g groupname|#gid] [-p prompt] [-U user
            name] [-u user name|#uid] [-g groupname|#gid] [command]
usage: sudo [-AbEHknPS] [-r role] [-t type] [-C fd] [-D level] [-g
            groupname|#gid] [-p prompt] [-u user name|#uid] [-g groupname|#gid]
            [VAR=value] [-i|-s] [<command>]
usage: sudo -e [-AknS] [-r role] [-t type] [-C fd] [-D level] [-g
            groupname|#gid] [-p prompt] [-u user name|#uid] file ...


FATAL: all hosts have already failed -- aborting

PLAY RECAP ******************************************************************** 
           to retry, use: --limit @/var/tmp/ansible/pb_complex.retry

test                       : ok=2    changed=1    unreachable=1    failed=0   
```

TL;DR : only variables defined in the playbook are expanded.

So this is definitively an issue (running from checkout).

Thanks.

**Comment by gildegoma** [comment]
+1 to solve this problem

**Comment by mpdehaan** [comment]
closing per above, thanks
